File: C++/remove_covered_intervals.cc
// https://leetcode.com/problems/remove-covered-intervals/

class Solution {
public:
    int removeCoveredIntervals(vector<vector<int>>& intervals) {
        
        sort(intervals.begin(), intervals.end(), myFn);
        int end, prev_end = 0;
        int count = 0;
        for(auto curr: intervals) {
            end = curr[1];
            if(prev_end < end){
                ++count;
                prev_end = end;
            }
        }
        return count;
    }
    static bool myFn(vector<int> &a, vector<int> &b){
            return a[0]==b[0] ? a[1]>=b[1]: a[0] < b[0];
    }
};

File: Python/Remove_Covered_Intervals.py
# Leetcode problem link: https://leetcode.com/problems/remove-covered-intervals/

class Solution:
    def removeCoveredIntervals(self, intervals: List[List[int]]) -> int:
        # sort intervals based on the start of interval
        # in case of same first element, list with greater end element should come first
        # eg [1,5], [1,17] -> [1,17], [1,5]
        intervals.sort(key=lambda x: [x[0], -x[1]])
        answer = len(intervals)
        maxEnd = 0 # keeping track of max end of interval
        for interval in intervals: 
            if interval[1] <= maxEnd:
                answer -= 1
            maxEnd = max(maxEnd, interval[1])
        return answer
